compositor_render/src/transformations/web_renderer/shared_memory.rs

```diff
@@ -21,53 +23,92 @@ impl SharedMemory {
         let path = root_path.join(node_id.to_string());
         Self::init_shared_memory_folder(&path)?;
 
-        Self::from_path(path.join(source_idx.to_string()), size)
+        let shmem_path = path.join(source_idx.to_string());
+        let mut shared_memory = Self {
+            state: SharedMemoryState::Unloaded,
+            path: shmem_path,
+            node_id: node_id.clone(),
+            source_idx,
+        };
+        shared_memory.init_shared_memory(size)?;
+
+        Ok(shared_memory)
     }
 
-    pub fn from_path(path: PathBuf, size: usize) -> Result<Self, SharedMemoryError> {
+    fn init_shared_memory(&mut self, size: usize) -> Result<(), SharedMemoryError> {
         let inner = ShmemConf::new()
-            .flink(&path)
+            .flink(&self.path)
             .size(size)
             .force_create_flink()
             .create()?;
+        self.state = SharedMemoryState::Loaded(inner);
+        Ok(())
+    }
 
-        Ok(Self {
-            inner: Some(inner),
-            path,
-        })
+    pub fn len(&self) -> usize {
+        self.state
+            .shared_memory()
+            .map(Shmem::len)
+            .unwrap_or_default()
     }
 
     pub fn to_path_string(&self) -> String {
         self.path.display().to_string()
     }
 
-    pub fn ensure_size(
+    pub fn request_resize(
         &mut self,
         size: usize,
         frame: &cef::Frame,
     ) -> Result<(), SharedMemoryError> {
-        let shmem_len = self.inner.as_ref().map(|shmem| shmem.len()).unwrap();
-        if shmem_len == size {
+        if let SharedMemoryState::PendingResize { new_size, .. } = &mut self.state {
+            *new_size = size;
             return Ok(());
         }
 
-        // TODO: This should be synchronised
-        let mut process_message = cef::ProcessMessage::new(UNEMBED_SOURCE_FRAMES_MESSAGE);
-        process_message.write_string(0, self.path.display().to_string());
-        frame.send_process_message(cef::ProcessId::Renderer, process_message)?;
-        // -----
+        let shared_memory = match self.state.unload() {
+            SharedMemoryState::Loaded(shared_memory)
+            | SharedMemoryState::PendingResize { shared_memory, .. } => shared_memory,
+            SharedMemoryState::Unloaded => return Err(SharedMemoryError::SharedMemoryInaccessible),
+        };
+        self.state = SharedMemoryState::PendingResize {
+            shared_memory,
+            new_size: size,
+        };
+
+        let mut msg = cef::ProcessMessage::new(UNEMBED_SOURCE_MESSAGE);
+        msg.write_string(0, self.to_path_string());
+        msg.write_string(1, self.node_id.to_string());
+        msg.write_int(2, self.source_idx as i32);
+        frame.send_process_message(cef::ProcessId::Renderer, msg)?;
+
+        Ok(())
+    }
 
+    pub fn resolve_pending_resize(&mut self) -> Result<(), SharedMemoryError> {
```

Comment:
resolve is very confusing here, if I understand correctly this function is "resizing" shared memory  if current state is already in pending resize state. Is that correct?

- You are mixing here logic on how to resize with when to resize.
- Name `resolve` is not a correct word here `finalize_pending_resize` or `execute_pending_resize` would be more appropriate.
- I might be missing sth important here, but is the fact that there is a resize pending sth that should even be part of the shared memory abstraction?